samza-api/src/main/java/org/apache/samza/storage/kv/KeyValueStore.java

```diff
@@ -141,4 +144,9 @@ default void deleteAll(List<K> keys) {
    * Flushes this key-value store, if applicable.
    */
   void flush();
+
+  /**
+   * Checkpoint the store snapshot.
+   */
+  Optional<Path> checkpoint(String id);
```

Comment:
1. Does absence of `Path` here denote a no-op / not-implemented checkpoint or could it be the case where implementers choose to return empty optional on some failure scenario? Does that distinction matter upstream?
2. Should we consider returning a complex object w/ status, result and exception since this might be running on a separate thread (for async commit) and exception handling and bubbling exception will be easier and callers can make that choice. what do you think?

Reply:
1. If None, changelog offsets will not be written to the store (e.g. for persistent stores that don't support checkpoints) (PR #1164 , TransactionalStateTaskStorageManager). If exception, container will shut down. I'll update the javadocs in a follow up documentation PR after #1164. I'll also add a test in #1164 to ensure that any exception creating store checkpoint is propagated in TransactionalStateTaskStorageManager. There's already a test that ensure any exception thrown in TaskInstance#commit causes is propagated.

2. Transactional state is currently not supported for async commits. I'll mark this method as unstable so that we can evolve it if necessary. The behavior is: if checkpoints supported, return a path, else return None. On errors, throw an exception. Not sure if separate status, result and exception will help here?

Reply:
Ideally we'd make this a default method with the default impl returning an empty optional. However because of the way store wrapping works (cached store, logged store etc.), the correct default behavior for wrapper stores is to call the underlying store. Given the risk of accidentally forgetting to call the underlying store checkpoint in any new wrapper implementations if we make this a default method, I think it'll be better to keep this as a new required method.

Reply:
Sure, please you make sure we throw an exception if transactional state is enabled w/ async commits.

Reply:
Created SAMZA-2336 to track.